File: qt/multimedia/gen_qvideoframeinput.cpp
// +build ignore

#include <QMediaCaptureSession>
#include <QMetaObject>
#include <QObject>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <QVideoFrame>
#include <QVideoFrameFormat>
#include <QVideoFrameInput>
#include <qvideoframeinput.h>
#include "gen_qvideoframeinput.h"
class MiqtVirtualQVideoFrameInput : public virtual QVideoFrameInput {
public:
MiqtVirtualQVideoFrameInput(): QVideoFrameInput() {};
MiqtVirtualQVideoFrameInput(const QVideoFrameFormat& format): QVideoFrameInput(format) {};
MiqtVirtualQVideoFrameInput(QObject* parent): QVideoFrameInput(parent) {};
MiqtVirtualQVideoFrameInput(const QVideoFrameFormat& format, QObject* parent): QVideoFrameInput(format, parent) {};

virtual ~MiqtVirtualQVideoFrameInput() = default;
// cgo.Handle value for overwritten implementation
	intptr_t handle__MetaObject = 0;
// Subclass to allow providing a Go implementation
	virtual const QMetaObject* metaObject() const override {
if (handle__MetaObject == 0) {
return QVideoFrameInput::metaObject();
}
QMetaObject* callback_return_value = miqt_exec_callback_QVideoFrameInput_MetaObject(const_cast<MiqtVirtualQVideoFrameInput*>(this), handle__MetaObject);

		return callback_return_value;
	}
// Wrapper to allow calling protected method
	QMetaObject* virtualbase_MetaObject() const {

		return (QMetaObject*) QVideoFrameInput::metaObject();

	}
// cgo.Handle value for overwritten implementation
	intptr_t handle__Metacast = 0;
// Subclass to allow providing a Go implementation
	virtual void* qt_metacast(const char* param1) override {
if (handle__Metacast == 0) {
return QVideoFrameInput::qt_metacast(param1);
}
const char* sigval1 = (const char*) param1;

		void* callback_return_value = miqt_exec_callback_QVideoFrameInput_Metacast(this, handle__Metacast, sigval1);

		return callback_return_value;
	}
// Wrapper to allow calling protected method
	void* virtualbase_Metacast(const char* param1) {

		return QVideoFrameInput::qt_metacast(param1);

	}
};
QVideoFrameInput* QVideoFrameInput_new() {
return new MiqtVirtualQVideoFrameInput();
}
QVideoFrameInput* QVideoFrameInput_new2(QVideoFrameFormat* format) {
return new MiqtVirtualQVideoFrameInput(*format);
}
QVideoFrameInput* QVideoFrameInput_new3(QObject* parent) {
return new MiqtVirtualQVideoFrameInput(parent);
}
QVideoFrameInput* QVideoFrameInput_new4(QVideoFrameFormat* format, QObject* parent) {
return new MiqtVirtualQVideoFrameInput(*format, parent);
}
void QVideoFrameInput_virtbase(QVideoFrameInput* src
, QObject** outptr_QObject
) {
*outptr_QObject = static_cast<QObject*>(src);
}
QMetaObject* QVideoFrameInput_MetaObject(const QVideoFrameInput* self) {
	return (QMetaObject*) self->metaObject();
}
void* QVideoFrameInput_Metacast(QVideoFrameInput* self, const char* param1) {
	return self->qt_metacast(param1);
}
struct miqt_string QVideoFrameInput_Tr(const char* s) {
	QString _ret = QVideoFrameInput::tr(s);
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
bool QVideoFrameInput_SendVideoFrame(QVideoFrameInput* self, QVideoFrame* frame) {
	return self->sendVideoFrame(*frame);
}
QVideoFrameFormat* QVideoFrameInput_Format(const QVideoFrameInput* self) {
	return new QVideoFrameFormat(self->format());
}
QMediaCaptureSession* QVideoFrameInput_CaptureSession(const QVideoFrameInput* self) {
	return self->captureSession();
}
void QVideoFrameInput_ReadyToSendVideoFrame(QVideoFrameInput* self) {
	self->readyToSendVideoFrame();
}
void QVideoFrameInput_connect_ReadyToSendVideoFrame(QVideoFrameInput* self, intptr_t slot) {
	MiqtVirtualQVideoFrameInput::connect(self, static_cast<void (QVideoFrameInput::*)()>(&QVideoFrameInput::readyToSendVideoFrame), self, [=]() {
		miqt_exec_callback_QVideoFrameInput_ReadyToSendVideoFrame(slot);
	});
}
struct miqt_string QVideoFrameInput_Tr2(const char* s, const char* c) {
	QString _ret = QVideoFrameInput::tr(s, c);
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
struct miqt_string QVideoFrameInput_Tr3(const char* s, const char* c, int n) {
	QString _ret = QVideoFrameInput::tr(s, c, static_cast<int>(n));
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
void QVideoFrameInput_override_virtual_MetaObject(void* self, intptr_t slot) {
	dynamic_cast<MiqtVirtualQVideoFrameInput*>( (QVideoFrameInput*)(self) )->handle__MetaObject = slot;
}
QMetaObject* QVideoFrameInput_virtualbase_MetaObject(const void* self) {
	return ( (const MiqtVirtualQVideoFrameInput*)(self) )->virtualbase_MetaObject();
}
void QVideoFrameInput_override_virtual_Metacast(void* self, intptr_t slot) {
	dynamic_cast<MiqtVirtualQVideoFrameInput*>( (QVideoFrameInput*)(self) )->handle__Metacast = slot;
}
void* QVideoFrameInput_virtualbase_Metacast(void* self, const char* param1) {
	return ( (MiqtVirtualQVideoFrameInput*)(self) )->virtualbase_Metacast(param1);
}
void QVideoFrameInput_Delete(QVideoFrameInput* self, bool isSubclass) {
	if (isSubclass) {
		delete dynamic_cast<MiqtVirtualQVideoFrameInput*>( self );
	} else {
		delete self;
	}
}

File: qt/multimedia/gen_qvideoframeinput.go
package multimedia

import (
	"unsafe"

	"github.com/mappu/miqt/qt"
)

type QVideoFrameInput struct {
	h          uintptr
	isSubclass bool
}

// NewQVideoFrameInput constructs a new QVideoFrameInput object.
func NewQVideoFrameInput() *QVideoFrameInput {
	g := newQVideoFrameInput(QVideoFrameInput_new())
	g.isSubclass = true
	return g
}

// NewQVideoFrameInput2 constructs a new QVideoFrameInput object.
func NewQVideoFrameInput2(format *QVideoFrameFormat) *QVideoFrameInput {
	g := newQVideoFrameInput(QVideoFrameInput_new2(format.cPointer()))
	g.isSubclass = true
	return g
}

// NewQVideoFrameInput3 constructs a new QVideoFrameInput object.
func NewQVideoFrameInput3(parent *qt.QObject) *QVideoFrameInput {
	g := newQVideoFrameInput(QVideoFrameInput_new3((*QObject)(parent.UnsafePointer())))
	g.isSubclass = true
	return g
}

// NewQVideoFrameInput4 constructs a new QVideoFrameInput object.
func NewQVideoFrameInput4(format *QVideoFrameFormat, parent *qt.QObject) *QVideoFrameInput {
	g := newQVideoFrameInput(QVideoFrameInput_new4(format.cPointer(), (*QObject)(parent.UnsafePointer())))
	g.isSubclass = true
	return g
}

func (this *QVideoFrameInput) MetaObject() *qt.QMetaObject {
	return qt.UnsafeNewQMetaObject(unsafe.Pointer(QVideoFrameInput_MetaObject(this.h)))
}

func (this *QVideoFrameInput) Metacast(param1 string) unsafe.Pointer {
	param1_Cstring := CString(param1)
	defer free(unsafe.Pointer(param1_Cstring))
	return (unsafe.Pointer)(QVideoFrameInput_Metacast(this.h, param1_Cstring))
}

func QVideoFrameInput_Tr(s string) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	var _ms struct_miqt_string = QVideoFrameInput_Tr(s_Cstring)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QVideoFrameInput) SendVideoFrame(frame *QVideoFrame) bool {
	return (bool)(QVideoFrameInput_SendVideoFrame(this.h, frame.cPointer()))
}

func (this *QVideoFrameInput) Format() *QVideoFrameFormat {
	_goptr := newQVideoFrameFormat(QVideoFrameInput_Format(this.h))
	_goptr.GoGC() // Qt uses pass-by-value semantics for this type. Mimic with finalizer
	return _goptr
}

func (this *QVideoFrameInput) CaptureSession() *QMediaCaptureSession {
	return newQMediaCaptureSession(QVideoFrameInput_CaptureSession(this.h))
}

func (this *QVideoFrameInput) ReadyToSendVideoFrame() {
	QVideoFrameInput_ReadyToSendVideoFrame(this.h)
}

func (this *QVideoFrameInput) OnReadyToSendVideoFrame(slot func()) {
	QVideoFrameInput_connect_ReadyToSendVideoFrame(this.h, intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QVideoFrameInput_ReadyToSendVideoFrame
func miqt_exec_callback_QVideoFrameInput_ReadyToSendVideoFrame(cb intptr_t) {
	gofunc, ok := cgo.Handle(cb).Value().(func())
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	gofunc()
}

func QVideoFrameInput_Tr2(s string, c string) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	c_Cstring := CString(c)
	defer free(unsafe.Pointer(c_Cstring))
	var _ms struct_miqt_string = QVideoFrameInput_Tr2(s_Cstring, c_Cstring)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func QVideoFrameInput_Tr3(s string, c string, n int) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	c_Cstring := CString(c)
	defer free(unsafe.Pointer(c_Cstring))
	var _ms struct_miqt_string = QVideoFrameInput_Tr3(s_Cstring, c_Cstring, (int)(n))
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QVideoFrameInput) callVirtualBase_MetaObject() *qt.QMetaObject {
	return qt.UnsafeNewQMetaObject(unsafe.Pointer(QVideoFrameInput_virtualbase_MetaObject(unsafe.Pointer(this.h))))
}

func (this *QVideoFrameInput) OnMetaObject(slot func(super func() *qt.QMetaObject) *qt.QMetaObject) {
	if !this.isSubclass {
		panic("miqt: can only override virtual methods for directly constructed types")
	}
	QVideoFrameInput_override_virtual_MetaObject(unsafe.Pointer(this.h), intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QVideoFrameInput_MetaObject
func miqt_exec_callback_QVideoFrameInput_MetaObject(self QVideoFrameInput, cb intptr_t) *QMetaObject {
	gofunc, ok := cgo.Handle(cb).Value().(func(super func() *qt.QMetaObject) *qt.QMetaObject)
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	virtualReturn := gofunc((&QVideoFrameInput{h: self}).callVirtualBase_MetaObject)

	return (*QMetaObject)(virtualReturn.UnsafePointer())
}

func (this *QVideoFrameInput) callVirtualBase_Metacast(param1 string) unsafe.Pointer {
	param1_Cstring := CString(param1)
	defer free(unsafe.Pointer(param1_Cstring))

	return (unsafe.Pointer)(QVideoFrameInput_virtualbase_Metacast(unsafe.Pointer(this.h), param1_Cstring))
}

func (this *QVideoFrameInput) OnMetacast(slot func(super func(param1 string) unsafe.Pointer, param1 string) unsafe.Pointer) {
	if !this.isSubclass {
		panic("miqt: can only override virtual methods for directly constructed types")
	}
	QVideoFrameInput_override_virtual_Metacast(unsafe.Pointer(this.h), intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QVideoFrameInput_Metacast
func miqt_exec_callback_QVideoFrameInput_Metacast(self QVideoFrameInput, cb intptr_t, param1 *const_char) unsafe.Pointer {
	gofunc, ok := cgo.Handle(cb).Value().(func(super func(param1 string) unsafe.Pointer, param1 string) unsafe.Pointer)
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}
	// Convert all CABI parameters to Go parameters
	param1_ret := param1
	slotval1 := GoString(param1_ret)
	virtualReturn := gofunc((&QVideoFrameInput{h: self}).callVirtualBase_Metacast, slotval1)

	return virtualReturn
}
